lib/nectar/worker.ex

```diff
@@ -3,49 +3,75 @@ defmodule Nectar.Worker do
 
   require Logger
 
+  def init(args) do
+    {:ok, args}
+  end
+
   def start_link(socket) do
     Task.start_link(fn ->
-        {:ok, client} = :gen_tcp.accept(socket)
-        Logger.debug "Client Connected"
-        serve(client)
+      {:ok, client} = :gen_tcp.accept(socket)
+      Logger.debug "client connected"
+      serve(client)
     end)
   end
 
   defp serve(client) do
     client
-    |> read_line()
-    |> write_line(client)
+    |> read_request()
+    |> write_response()
+  end
+
+  defp read_request(client), do: read_request_line(client)
+
+  defp read_request_line(client) do
+    Logger.debug("in read_request_line/1")
+    request_line = read_line(client)
+    read_header(client, %{request_line: request_line, headers: []})
   end
 
-  defp read_line(client) do
-    with {:ok, data} <- :gen_tcp.recv(client, 0) do
-      Logger.debug(data, label: "data")
-      data
+  defp read_header(client, %{headers: headers} = request) do
+    Logger.debug("in read_header/2")
+    case read_line(client) do
+      "" -> read_message_body(client, request)
+      header_line when is_binary(header_line) -> read_header(client, %{request | headers: [header_line] ++ headers})
+      other ->
+        Logger.error("read_line/1 returned #{inspect other}")
+        {:client, :error, other}
     end
   end
 
-  defp write_line(_line, client) do
-    message = "Hello, World!"
+  defp read_message_body(client, request) do
+    Logger.debug("in read_message_body/2")
+    {client, request}
+  end
 
-    response = """
-    HTTP/1.1 200 OK
+  # FIXME: Doesn't read a line -- reads a chunk of data. Should break it up into lines.
+  defp read_line(client), do: with {:ok, data} <- :gen_tcp.recv(client, 0), do: String.trim_trailing(data)
+
+  defp write_response({client, :error, reason}), do: build_response(500, "Internal Server Error", (inspect reason)) |> send_response(client)
+  defp write_response({client, _request}), do: build_response(200, "OK", "Hello, world!") |> send_response(client)
```

Comment:
Just a style thing, If there is more than one operation after the `do` I like to see a legit block. Have you run this through `mix format`? If the code formatter accepts it then I'm ok with leaving it.